Adding test setup class that can override how a SUT is built.

appakz opened the pull request:
This may still need some work, but here's an initial implementation for an idea I've been kicking around.

The idea here is to provide a way to control how the SUT is built by providing a Func<TUnit> that can be used to build one manually. The lifecycle of how an SUT is managed doesn't change, this just short-circuits using the auto mocking container and will invoke the provided function instead if provided.

This is how you use it:

```c#
public void RunTests()
{
  When("The SUT does something intesresting", () => SUT.DoSomethingInteresting())
  .BuildSUTUsing(() => new SUTImplementation(new Dependency1()));

  //Givens and Thens go here
}
```

Thoughts?

@rockhymas @Fammy @ronrat @mmertsock ?

ronrat commented:
Here's another possible approach:

    public void RunTests()
    {
        With(() => new SUTImplementation(new Dependency1()))
        .When("The SUT does something intesresting", () => SUT.DoSomethingInteresting());

      //Givens and Thens go here
    }

I think it reads a little easier. Build the SUT and then use it vs. showing the use of it before defining how it is built. Of course, leaving out the `With()` will result in use the default auto-mocking.

There might be a better name than `With()`. Another option is `WithSUT()`.

rockhymas commented:
I really like the "With" approach, and I think it's a great name. For cases where we want to explicitly test construction, maybe we have a new method called WhenConstructing, which just calls the With clause (or the default auto-mocking). It could be used either off the With, as above, or when the With is left out.

appakz commented:
That's an interesting idea... I'll take a stab at implementing it that way and see how it shakes out.

mmertsock commented:
IMO putting the `When` first is a little more in the spirit of BDD, in that the use case is front and center, and the SUT construction tends to be an implementation detail. But I don't hold that opinion strongly, I could be convinced with either approach. Anyway I like the general idea of this feature, it will be useful.

ronrat commented:
I see your point, @mmertsock, about the ordering and I wouldn't be opposed to that. It reads like "When testing With", which makes sense.

jsternal commented:
What about something like this?

`Given("a concrete dependency", new ConstructorArgument("dependency", new Dependency());`

I found myself wishing for something similar this week, though my situation was a little different: I wanted to specify one constructor argument (an primitive invariant - in this case a string) and let SpecEasy handle the rest using the auto-mocking. Today, given an SUT with the constructor `SUT(IDependency dependency, string invariant)`, I have to do this:

    private IDependency dependency; // I'd usually auto-mock this

    override BeforeEachExample()
    {
         dependency = Mock<IDependency>();
    }

    public void TheSpec()
    {
        When("doing something", SUT.DoSomething());
        Given("my invariant is empty", SUT = new SUT(dependency, string.Empty)).Verify(() =>
        {
            // Specs here
     ...});
    }

In which case this would be the alternative:

    public void TheSpec()
    {
        When("doing something", SUT.DoSomething());
        Given("my invariant is empty", new ConstructorArgument("invariant", string.Empty)).Verify(() =>
        {
            // Specs here
     ...});
    }

rockhymas commented:
Ooh, I like the possibilities here, @jsternal. I'm pretty sure it would be quite easy to make any needed modifications to speceasy/tinyioc to handle these named constructor arguments. It might not solve the entire need here, but it's a good first step, and it's one I think we should have, even if we eventually do something further using one of the other suggestions.

dbertram commented:
I've been brainstorming about this (and some related pain points) and was thinking an alternate approach might address some (most?) of the desires here in a more general way:

- modify Spec's `protected TUnit SUT` property so that it's `virtual`

_or_
- modify Spec's `protected TUnit SUT` so instead of returning `Get<TUnit>()` it defers to a `protected virtual TUnit ConstructSUT()` that's called if SUT hasn't already been constructed

So your Spec might end up looking something like this:

```c#
public class MySpec : Spec<Foo>
{
  public void MethodName()
  {
    When("doing something", () => SUT.DoSomething());
    ...
  }

  protected override Foo ConstructSUT()
  {
    return new Foo(whatever, theOtherThing, etc);
  }
}
```

**Pros** (imho):
- useful when your spec methods need to stub different values for constructor arguments for different test cases (i.e., assigning SUT from `BeforeEachExample` isn't workable)
- useful when updating existing specs that rely on SUT being lazily constructed without having to determine _when_ SUT will be constructed by looking at the control flow of the spec (particularly useful when updating Specs that are testing event handling, in my experience)
- easy deferral to the base implementation (e.g. `return base.ConstructSUT();`)
- less repetition if you _always_ need to customize how SUT is constructed vs. the `With` or `BuildSUTUsing` syntaxes suggested above

**Cons**:
- potentially trickier to customize per-Spec method than something like the suggested `With` or `BuildSUTUsing` syntaxes suggested above (could use class fields that are assigned in `Given`s to mitigate potentially)
- less automatic/succinct than the `Given(string description, ConstructorArgument arg)` overload suggested above

What do you think?

rockhymas commented:
Given that the Spec<T> class is testing a single type T, I think it makes sense to use something at the spec class level (i.e. a protected virtual method) to specify how the type T is constructed, so I like that part of your plan. I think your first con is fully mitigated by using member fields for handling constructor params, or using the ConstructorArguments formulation proposed by @jsternal.

As for the second con, I see this approach and a way to specify constructor arguments as orthogonal, and both have value. They can complement each other. In some cases you're always going to use the same constructor, so just overriding `ConstructSUT` will totally solve your problem. In other cases you want to specify constructor args in your test contexts, and having that functionality makes your life easier. One other advantage of ContructorArguments is that they will work when testing abstract classes, and ConstructSUT won't*. 

I could see using them together as well:

```
public class MySpec : Spec<Foo>
{
  public void MethodName()
  {
    When("doing something", () => SUT.DoSomething());
    ...
  }

  protected override Foo ConstructSUT()
  {
    ConstructorArgument("invariant", string.Empty);
    return base.ConstructSUT();
  }
}
```

* Actually, you could use ConstructSUT to go with a totally different mocking framework, though there would be other changes we'd want to make, like letting the user override the FallbackResolutionAction for when TinyIoC needs a type it cannot construct. But that's a separate issue.

appakz commented:
I like the `virtual ConstructSUT()` method approach; it's simple and we've already established a pattern of exposing `virtual` methods on the base spec for making things more powerful/flexible (e.g. `BeforeEachExample`).

ronrat commented:
If I'm understanding the `virtual ConstructSUT()` approach, I could do something like:

    public class MySpec : Spec<Foo>
    {
        string theOtherThing = null;
        
        public void MethodName()
        {
            When("doing something", () => SUT.DoSomething());

            Given("the other thing is empty", () => theOtherThing = string.Empty)).Verify(() =>
            {
                // Specs here
            });
        }

        protected override Foo ConstructSUT()
        {
            var whatever = Get<Whatever>();
            return new Foo(whatever, theOtherThing);
        }
    }

I really like that. It reads and looks like what, IMO, a Spec should. It's also easy to see in my Spec class where `theOtherThing` is used. Making both this and the `ConstructorArgument` approach available would get my vote, as it gives the developer a choice of styles based upon preference. Using `ConstructorArgument` seems a bit more clunky but I can see myself using that at times when I'm lazy and don't want to override the `ConstructSUT`. Based upon the `ConstructorArgument` class name, it's still clear what's going on underneath.

jsternal commented:
Really nice, I think it's a simple clear design that seems like it already ought to exist.

This may make the `Given(string description, ConstructorArgument arg)` proposal unnecessary, given that we can manipulate class fields to achieve the same thing. At first glance, that dismayed me a bit, because that style (manipulating class fields to set up specs) isn't entirely intuitive (for newcomers). But it's central to using SpecEasy effectively, so I got over that quickly.

appakz commented:
@rockhymas @jsternal 

@dbertram and I got this cleaned up a bit and I think it's good to go now if you all want to take another look. To summarize what we did here:

- Added a `protected virtual T ConstructSUT()` method that will defer to our TinyIoC auto mocking container to construct an SUT, but can be overridden by specs if needed.
- Added some specs to demonstrate how overriding the construction of an SUT works.

dbertram commented on SpecEasy/GenericSpec.cs:
```diff
@@ -107,12 +133,12 @@ internal override void BeforeEachInit()
         {
-            get { return Get<TUnit>(); }
+            get { return GetSUTInstance(); }
             set { Set(value); }
```
Should we be updating `alreadyConstructedSUT` and `constructedSUTInstance` via the setter here?

Otherwise, if someone assigns explicitly to `SUT`, we could still trigger construction, no?

appakz replied:
Good question; we probably *should* do that, but now I"m wondering if exposing the setter here was just a band aid over the fact that we previously had no way to control how an SUT was constructed.

Removing the `set` here would be a breaking API change, but maybe we pull this out unless there's a valid scenario where you still need it even with the `ConstructSUT` approach.

dbertram replied:
I don't know if it was a band aid, or maybe just keeping in line with the `Get<T>`/`Set<T>` convention used elsewhere with the mocking container?

appakz replied:
I'm not sure either, but we have exposed a potentially confusing scenario.

- If you never override `ConstructSUT` or assign to the `SUT` property, we just auto mock SUT.
- If you never override `ConstructSUT`, but you do assign to the `SUT` property, then we will use whatever you assigned.
- If you override `ConstructSUT` and you never assign to the `SUT` property, then we just use `ConstructSUT` 's result.
- *but* if you both override `ConstructSUT` ***and*** set the `SUT` property then we'll never use what you assign because `GetSUTInstance` will invoke `ConstructSUT` and then cache the result in `constructedSUTInstance` and provide it to the mocking container).

Because of that last scenario this sort of feels like how we used to let you shoot yourself in the foot by making  you remember that you had to call `base.BeforeEachExample` whenever you did an override.

We could try to prevent this without removing the `set` by checking to see if `alreadyConstructedSUT` is `true` when you invoke the setter, but that won't catch scenarios where you set the `SUT` before you access it (which is the only way it works now).

Alternatively, we could allow a set of `SUT` to trump anything that might be returned by `ConstructSUT`, but that feels as awkward and arbitrary as letting `ConstructSUT` override what you set in `SUT`.

dbertram replied:
I think the simplest solution that doesn't introduce a breaking change is just to set `alreadyConstructedSUT` and `constructedSUTInstance` in SUT's setter.

Then the workflow for "How is SUT constructed?" works roughly like this:

- No customization => automock
- Override `CreateSUT` => use whatever that returns
- Assign explicitly to `SUT` => use whatever you assign
- Override `CreateSUT` _and_ assign explicitly to `SUT` => Use the explicitly assigned value after it's been assigned, otherwise get whatever `CreateSUT` returns (as laziness dictates).

That last point is a little tricky, but kind of lines up with how SpecEasy already works: if you assign to SUT, you'll get that. If you don't, SpecEasy will try to create it for you (via automock). This new functionality is adding an extensibility point in the "we don't have SUT yet, but it's being requested" workflow.

Of course, at this point I can't claim to be unbiased / looking at this through the eyes of a "new" SpecEasy user. :/

We _could_ drop the setter on SUT, but I still feel like there are use cases for wanting to explicitly assign to it. This is maybe a poor example, but if you had a class with two constructors, it's quick and easy to test each one via `Given`s if you can assign to SUT:

```c#
Given("SUT is constructed using constructor A", () => SUT = new FooBar(a)).Verify(() => ...);
Given("SUT is constructed using constructor B", () => SUT = new FooBar(b)).Verify(() => ...);
```

If we _remove_ the setter on SUT, you'd have to use state in your spec to accomplish that...probably something like this:

```c#
class MySpec : Spec<FooBar>
{
    private bool useContructorA = false;
    ...

    public void Constructor()
    {
        Given("we use constructor A", () => useConstructorA = true).Verify(() => ...);
        Given("we use constructor B", () => useConstructorA = false).Verify(() => ...);
    }

    protected override FooBar ConstructSUT()
    {
        return useContructorA ? new FooBar(a) : new FooBar(b);
    }
}
```

Or similar logic in `BeforeEachExample`

appakz replied:
OK I agree that keeping the setter for `SUT` is probably the lowest friction way to go. I'll update that and add a spec for it.

dbertram commented on SpecEasy/GenericSpec.cs:
```diff
@@ -88,6 +113,7 @@ internal override void BeforeEachInit()
         {
             base.BeforeEachInit();
 
+            alreadyConstructedSUT = false;
```
Another minor note, but should be setting `constructedSUTInstance` to `null` here for clarity?

appakz replied:
We can, but it'll have to be set to `default(TUnit)` because we don't have a `class` generic constraint in place (though we maybe should?)

appakz commented:
@dbertram see fca321a 

This updates our cached SUT instance when invoking the setter which gives us the behavior we had discussed previously.

dbertram commented:
@appakz Looks good!
